integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonAnalysisRules.scala

```diff
@@ -52,7 +53,26 @@ object CarbonPreInsertionCasts extends Rule[LogicalPlan] {
         )
     }
     if (child.output.size >= relation.carbonRelation.output.size) {
-      InsertIntoCarbonTable(relation, p.partition, p.child, p.overwrite, p.ifNotExists)
+      var index = 0
+      val newChildOutput = child.output.map { column =>
+        index = index + 1
+        column match {
+          case attr: Alias =>
+            Alias(attr.child, s"col$index")(attr.exprId)
+          case attr: Attribute =>
+            Alias(attr, s"col$index")(NamedExpression.newExprId)
+          case attr => attr
+        }
+      }
+      if (newChildOutput == child.output) {
+        InsertIntoCarbonTable(relation, p.partition, p.child, p.overwrite, p.ifNotExists)
+      } else {
+        InsertIntoCarbonTable(relation,
+          p.partition,
+          Project(newChildOutput, child),
```

Comment:
I think you can move the creation of `InsertIntoCarbonTable` below else clause, so that it will take care of both if and else case. Only one parameter is different.